crates/dt-workspace/src/config.rs

```diff
@@ -0,0 +1,151 @@
+    }
+
+    /// Loads config file from file
+    pub fn load(path: &Path) -> Result<Self, ConfigError> {
+        toml::from_str(&fs_err::read_to_string(path)?).map_err(ConfigError::Toml)
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum ConfigError {
+    #[error("Failed to read config")]
+    Io(#[from] std::io::Error),
+
+    #[error("Failed to deserialize config")]
+    Toml(#[from] toml::de::Error),
+}
+
+#[cfg(test)]
+mod tests {
+
+    use std::{env::current_dir, io};
+
+    use super::*;
+
+    #[test]
+    fn find_cwd() {
+        assert_eq!(
+            current_dir()
+                .unwrap()
+                .join(CONFIG_FILENAME)
+                .canonicalize()
+                .unwrap(),
+            Config::find_config(None).unwrap().canonicalize().unwrap()
+        );
+    }
+
+    #[test]
+    fn find_tests() {
+        assert_eq!(
+            current_dir()
+                .unwrap()
+                .join("test_data")
+                .join(CONFIG_FILENAME)
+                .canonicalize()
+                .unwrap(),
+            Config::find_config(Some(Path::new("test_data")))
+                .unwrap()
+                .canonicalize()
+                .unwrap()
+        );
+    }
+
+    #[test]
+    fn find_ancestor() {
+        assert_eq!(
+            current_dir()
+                .unwrap()
+                .join("test_data")
+                .join(CONFIG_FILENAME)
+                .canonicalize()
+                .unwrap(),
+            Config::find_config(Some(&Path::new("test_data").join("empty")))
+                .unwrap()
+                .canonicalize()
+                .unwrap()
+        );
+    }
+
+    #[test]
+    fn load_cwd_parent() {
+        let config = Config::load_from_parent(Path::new(".")).unwrap();
+        assert_eq!(vec![PathBuf::from("test_data")], config.include_paths);
+    }
+
+    #[test]
+    fn load_cwd() {
+        let config = Config::load(Path::new(CONFIG_FILENAME)).unwrap();
+        assert_eq!(vec![PathBuf::from("test_data")], config.include_paths);
+    }
+
+    #[test]
+    fn load_tests_parent() {
+        let config = Config::load_from_parent(Path::new("test_data")).unwrap();
+        assert_eq!(vec![PathBuf::from(".")], config.include_paths);
+    }
+
+    #[test]
+    fn load_tests() {
+        let config = Config::load(&Path::new("test_data").join(CONFIG_FILENAME)).unwrap();
+        assert_eq!(vec![PathBuf::from(".")], config.include_paths);
+    }
+
+    #[test]
+    #[should_panic]
+    fn load_panic() {
+        // invalid because we expect parent + CONFIG_FILENAME here
+        Config::load(Path::new("test_data")).unwrap();
+    }
+
+    #[test]
+    fn load_not_exists_parent() {
+        if let Err(ConfigError::Io(err)) = Config::load_from_parent(Path::new("not_exists")) {
+            if err.kind() == io::ErrorKind::NotFound {
+                return;
+            }
+        };
+
+        panic!("expected io::ErrorKind::NotFound")
+    }
```

Comment:
Sorry, didn't notice this before. Could you use `should_panic` with an expected substring?
E.g.

```suggestion
    #[test]
    #[should_panic = "not found"] // I don't remember the syntax
    fn load_not_exists_parent() {
      Config::load_from_parent(Path::new("not_exists")).unwrap(); // add let _ = .. as necessary
    }
```

Reply:
I believe it's bad idea, it expects whole string:
```
 panic message: `"called `Result::unwrap()` on an `Err` value: Io(Custom { kind: NotFound, error: Error { kind: OpenFile, source: Os { code: 2, kind: NotFound, message: \"No such file or directory\" }, path: \"not_exists/.dt-tools.toml\" } })"`,
 expected substring: `"Is a directory"`
 ```